Lombiq.Tests.UI/Services/GitHubAnnotationWriter.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Xunit.Abstractions;
+
+namespace Lombiq.Tests.UI.Services;
+
+public class GitHubAnnotationWriter
+{
+    private readonly ITestOutputHelper _testOutputHelper;
+
+    public GitHubAnnotationWriter(ITestOutputHelper testOutputHelper) => _testOutputHelper = testOutputHelper;
+
+    public void Annotate(LogLevel severity, string title, string message, string file, int line = 1)
+    {
+        ArgumentNullException.ThrowIfNull(message);
+        ArgumentNullException.ThrowIfNull(file);
+
+        if (file.Contains(',')) throw new ArgumentException("File name mustn't contain commas.", nameof(file));
+
+        title ??= severity.ToString();
+
+        var command = severity switch
+        {
+            LogLevel.Information => "notice",
+            LogLevel.Warning => "warning",
+            LogLevel.Error => "error",
+            LogLevel.Critical => "error",
+            _ => throw new ArgumentOutOfRangeException(
+                nameof(severity),
+                severity,
+                $"Only {nameof(LogLevel.Information)} - {nameof(LogLevel.Critical)} are valid."),
+        };
+
+        // We replace commas with reversed commas and double colons with the squared four dots character to avoid
+        // conflicts with the command parser. These are reasonably similar to carry the meaning, yet distinct enough to
+        // avoid misleading the reader. (For example if we replaced colons with "Armenian full stop" that looks
+        // identical, the user would have no idea why copying the output to a search yields no results when it should.)
+        title = title.Replace(',', '⹁').Replace("::", "⸬");
+
+        // Sanitize message:
+        message = message.Replace("\r", string.Empty).Replace("\n", " ");
+
+        _testOutputHelper.WriteLine(FormattableString.Invariant(
+            $"::{command} file={file},line={line},title={title}::{message}"));
+    }
+
+    public void ErrorInTest(Exception exception, ITestCase testCase)
```

Comment:
What Jocó has done under https://github.com/Lombiq/UI-Testing-Toolbox/pull/166 related to call stacks might be interesting here. Just food for thought.